### PR TITLE
SOLR-15342: Separate out a SolrJ-Zookeeper module

### PR DESCRIPTION
**Pull request by heythm**
https://issues.apache.org/jira/browse/SOLR-15342

I open this PR in draft to share the evolution of the work. I closed the old [PR](https://github.com/apache/solr/pull/584) since it has been open for a while, there have been quite a few changes since then in relation to ZooKeeper and some of the work has been contributed in a separate [PR](https://github.com/apache/solr/pull/749).
I resume the work done in the latter in this one

### CODE REVIEW

**Comment by HoustonPutman** [comment]
Thanks for getting this started again @haythemkh! As a first-pass I suggest that we put the current solrj functionality into `solrj/core`. Once that is committed, we can then do what this PR does, but split it out into `solrj/zookeeper` instead of `solrj-zookeeper`. I do not think we want to have all of the `solrj` libraries as folders under `solr/` directly.

**Comment by dsmiley** [comment]
> Once that is committed, we can then do what this PR does

@HoustonPutman A change like that risks Haythem having to start over an additional time.  I am encouraging him to get this done quickly so avoid it getting stale as what happened last time (which was depressing).  If we want to have solrj/core then I propose that happen in this issue at the end, or in a follow-on PR.

**Comment by HoustonPutman** [comment]
Ok I will do it following this getting merged

File: solr/solrj/src/java/org/apache/solr/client/solrj/io/sql/DatabaseMetaDataImpl.java [hunk 1]
```diff
@@ -118,7 +118,7 @@ public String getDatabaseProductVersion() throws SQLException {
     SolrClient solrClient = null;
     for (String node : liveNodes) {
       try {
-        String nodeURL = ZkStateReader.from(cloudSolrClient).getBaseUrlForNodeName(node);
+        String nodeURL = Utils.getBaseUrlForNodeName(node, /*getClusterProperty(URL_SCHEME, "http")*/ "http");
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/client/solrj/io/sql/DatabaseMetaDataImpl.java]: That's "nocommit" worthy :-)

I suppose CloudSolrClient.getClusterStateProvider should have a getBaseUrlForNodeName() instance method.  The ZK implementation could call the one on ZkStateReader so that this scheme is consulted.

File: solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/DeepRandomStream.java [hunk 2]
```diff
@@ -282,50 +277,6 @@ public List<TupleStream> children() {
     return solrStreams;
   }
 
-  public static Slice[] getSlices(
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/DeepRandomStream.java:285]: Where did this method go?

**Review by dsmiley — COMMENTED** [review]
Thanks for undertaking this large task!  I reviewed the commits; some of it was definitely deja vu from the first attempt that we did together.  I left some specific comments.  I need to check out the PR and see how it is for myself; I'm sure I'll catch some other things.

Overall comments, especially for anyone watching:
* The CloudHttp2SolrClient Builder show we need to use Java reflection to instantiate the ZK cluster state provider.  Ah well.
* Where do ZkStateReader constants go?  Maybe not necessarily in one place; maybe it's okay to have some copies in certain classes.  For example the client classes that set parameters; maybe that's one place as they are parameters after all.  But elsewhere in classes like DocCollection, Slice, etc. it's there because it's the field in the JSON of the cluster state which not coincidentally aligns with the parameter names.  That could be the second.  At least this makes sense to me.  I wouldn't undertake this refactoring in this issue but it'd be a nice follow-on.

File: solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/DeepRandomStream.java [hunk 3]
```diff
@@ -293,7 +293,7 @@ public static Slice[] getSlices(
 
     // check for alias or collection
 
-    List<String> allCollections = new ArrayList<>();
+    List<String> allCollections = new ArrayLigist<>();
```
[the remark this answers is not in the document]
Comment (haythemkh) [reply solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/DeepRandomStream.java]: This is already fixed in the PR

File: solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/DeepRandomStream.java (hunk 2 above)
Comment (haythemkh) [reply solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/DeepRandomStream.java:285]: I moved this to CloudSolrStream

File: solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/DeepRandomStream.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (haythemkh) [reply solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/DeepRandomStream.java]: Yes. This is already fixed in the PR :)

**Comment by cpoerschke** [comment]
oops, was trying to be helpful resolving the merge conflict but forgot to compile locally before pushing, sorry. will try to fix, or else commit reversion logically should also be possible? sorry for the noise.

**Comment by cpoerschke** [comment]
> oops, was trying to be helpful resolving the merge conflict but forgot to compile locally before pushing, sorry. will try to fix, or else commit reversion logically should also be possible? sorry for the noise.

added 3 commits to make it compile a little more but not fully. not sure if it was compiling before actually. again, sorry for the noise.

**Comment by dsmiley** [comment]
@haythemkh I know you are using the Google Java Format IntelliJ plugin but not the IntelliJ code style that accompanies it, which ensures that the import wildcards are configured correctly (the plugin can't do that part).  See https://github.com/google/google-java-format#intellij-android-studio-and-other-jetbrains-ides  

FYI @epugh it would be cool to mention this somewhere for contributor onboarding

**Comment by haythemkh** [comment]
I would love to have your feedback on what we want to do with the nocommits:
- These include copying ZkStateReader constants for solrj classes: [Utils](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/common/util/Utils.java#L89), [PerReplicaStates](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java#L182), [ClusterState](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/common/cloud/ClusterState.java#L49), [DocCollection](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java#L44), [Slice](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/common/cloud/Slice.java#L46), [Replica](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/common/cloud/Replica.java#L41), [ZkCoreNodeProps](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/common/cloud/ZkCoreNodeProps.java#L23), [CloudSolrClient](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudSolrClient.java#L93), [BaseHttpClusterStateProvider](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/client/solrj/impl/BaseHttpClusterStateProvider.java#L48), [CollectionAdminRequest](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/client/solrj/request/CollectionAdminRequest.java#L73), [CoreAdminRequest](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/client/solrj/request/CoreAdminRequest.java#L40)
- [Use raw "http" instead of going through #getClusterProperty in DatabaseMetaDataImpl](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/client/solrj/io/sql/DatabaseMetaDataImpl.java#L123)

**Comment by HoustonPutman** [comment]
> [PerReplicaStates](https://github.com/apache/solr/blob/521900d77ac938ba98a237fc1ca77700b168714d/solr/solrj/src/java/org/apache/solr/common/cloud/PerReplicaStates.java#L182)

Is there any reason for PRS not to be in the zk module? I don't think it would be used if we aren't communicating with ZK.


> Copy of ZkStateReader constants to solrj classes

Would it make more sense to have one `ClusterStateProps` class to hold these? We could also split it into `ClusterProps`, `CollectionProps`, etc, but that might be overkill.

**Comment by dsmiley** [comment]
PerReplicaStates is referenced in ClusterState, DocCollection, and Replica.  Based on my very limited understanding of PRS, it's a part of the state of a collection no matter how it's retrieved (via ZK or via SolrClient based ClusterStateProvider).  Am I right @noblepaul @chatman  ?

I do like your suggestion of those 3 "Props" classes for constants.  How do we get there from here?  Directly, even if it changes import statement in lots more classes (I would assume)?  This is the easiest path.  Or do we leave that for another JIRA issue (soon) and for now just remove the nocommits about this to make precommit happy?  Also relatively easy but leaves some WIP.  Or go part way here & now -- add the new classes & constants but only use them in solrj-zookeeper right now and then migrate the rest in another issue?  More work.

I would like to point out that this new separation did move some tests to solrj-zookeeper but it's only four -- that's it.  Many solrj tests indirectly use ZK and I suppose that's okay.  There is a compile time dependency on ZK from solrj tests that is unfortunate but couldn't be avoided in a few cases (without investing additional work):
* StreamingTest.streamLocalTests  calls `zkStateReader.forceUpdateCollection`
* TestCloudSolrClientConnections calls cluster.getZkClient()
* org.apache.solr.client.solrj.impl.HttpClusterStateSSLTest#testHttpClusterStateWithSSL calls cluster.getZkClient() 
Perhaps these might even move but on the other hand, it's not some big deal to add an explicit dependency on solrj tests to ZK when ZK is definitely already there at runtime via the solrj-zookeeper dependency.

There is a weird dependency check problem that I can't figure out:
```
Execution failed for task ':solr:solrj-zookeeper:analyzeTestClassesDependencies'.
> Dependency analysis found issues.
  usedUndeclaredArtifacts
   - org.apache.solr:solrj-zookeeper:10.0.0-SNAPSHOT@
```
So solrj-zookeeper's tests depend on solrj-zookeeper.  Well of course it does ;-)

**Comment by dsmiley** [comment]
A big obstacle I'm seeing is split packages between solrj & solrj-zookeeper with respect to javadocs.  For example the javadocs for solr-core will now contain broken links to most solrj-zookeeper classes because the javadoc tool can't figure this out (kind of a long-standing deficiency).  

Personally, I'm not very sympathetic to the needs of continuing to maintain a javadoc site.  I don't use them, I doubt others do, and moreover for open-source, really the source is what we want to see because it's better / more.  That's me, anyway.  I don't think maintaining them is more important than other goals like modularization, and thus I'm willing for there to be some degradation there.

Options:
* In our javadocs, stop **linking** out to solrj or solrj-zookeeper because they cannot be differentiated by the standard javadoc tool.
* Accept broken links; stop checking for them within javadoc folders.
* Fix it (not something I'm willing to invest in)

Of course in main/10.0 we can remove the split packages but that's a back-combat break that isn't okay for 9.x.

CC @uschindler -- I'm curious what you think

**Comment by janhoy** [comment]
> * In our javadocs, stop **linking** out to solrj or solrj-zookeeper because they cannot be differentiated by the standard javadoc tool.

This seems acceptable for 9.x, and in main branch we could change package names and re-introduce some links if needed?

**Comment by dsmiley** [comment]
@sigram There is an abstraction `SolrCloudManager` and some related classes (DistributedQueue, DistribStateManager, NodeStateProvider) that you introduced once upon a time that don't seem to have anything to do with SolrJ yet you put them there.  I believe these could go to Solr Core.  Most need to _at least_ go to Solrj-zookeeper because of DistributedQueue's reference of ZK classes.

**Comment by dsmiley** [comment]
For constants related to the collection's state, I put it in an inner interface DocCollection.CollectionSProps.  The "S" there is for State.  If that's removed, then it's possibly confused with CollectionProperties.  It's mostly an internal set of constants (it's not *parameters*) so I think it's okay it it doesn't roll off the tongue.  Likewise I put constants for the Replica in Replica.ReplicaSProps.  I'm not married to these choices.

There is a CollectionAdminParams class but it was missing quite a number of parameters that CollectionAdminRequest needed to specify so I moved them there.

There were a handful of other constants then I mostly made private and didn't fret over possible redundancy with ZkStateReader.

Overall I like having these props or params as constants in suitable locations, especially when we don't static import these constants.  It adds clarity to the spots where they are referred to as being related to parameters or properties or whatever it is.

The only thing that remains is the javadoc split package concern.  It should only be a handful of lines of code or so in render-javadoc.adoc to special case them to not inter-link.

**Comment by epugh** [comment]
At the risk of bikeshedding....   How about `ReplicaStateProps.LEADER` instead of `ReplicaSProps.LEADER` since we don't have an existing concept of `SProps` throughout our codebase?

File: gradle/documentation/render-javadoc.gradle [hunk 4]
```diff
@@ -141,13 +141,26 @@ configure(project(":solr:solrj")) {
   }
 }
 
+configure(project(":solr:solrj-zookeeper")) {
+  project.tasks.withType(RenderJavadocTask) {
+    // TODO: clean up split packages
```
Comment (epugh) [inline gradle/documentation/render-javadoc.gradle:146]: Could mention [SOLR-15604](https://issues.apache.org/jira/browse/SOLR-15604) here?

File: solr/solrj/src/java/org/apache/solr/common/cloud/CollectionStatePredicate.java [hunk 5]
```diff
@@ -18,14 +18,12 @@
...
 import java.util.Set;
-import java.util.concurrent.TimeUnit;
-import java.util.function.Predicate;
 
 /**
  * Interface to determine if a set of liveNodes and a collection's state matches some expecatations.
  *
- * @see ZkStateReader#waitForState(String, long, TimeUnit, CollectionStatePredicate)
- * @see ZkStateReader#waitForState(String, long, TimeUnit, Predicate)
+ * <p>{@code ZkStateReader#waitForState(String, long, TimeUnit, CollectionStatePredicate)} {@code
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/common/cloud/CollectionStatePredicate.java:25]: I see an open <p> but not a close one?   Should it be just two bullets instead?

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java [hunk 6]
```diff
@@ -68,7 +69,21 @@ protected CloudHttp2SolrClient(Builder builder) {
             "Both zkHost(s) & solrUrl(s) have been specified. Only specify one.");
       }
       if (builder.zkHosts != null) {
-        this.stateProvider = new ZkClientClusterStateProvider(builder.zkHosts, builder.zkChroot);
+        try {
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java:72]: Worth some docs on this?   Why do we need to use the Class.forName, is this potentially pluggable somehow?  Makes me think of JDBC drivers!

File: solr/solrj/build.gradle [hunk 7]
```diff
@@ -50,20 +41,15 @@ dependencies {
...
-  implementation ('org.apache.zookeeper:zookeeper-jute') {
-    exclude group: 'org.apache.yetus', module: 'audience-annotations'
-  }
-
   testImplementation project(':solr:test-framework')
   testImplementation project(':solr:core')
   testImplementation project(':solr:solrj')
   testRuntimeOnly project(':solr:modules:sql')
 
+  // ideally ZK centric tests move to solrj-zookeeper but sometimes we depend on ZK here anyway
```
Comment (epugh) [inline solr/solrj/build.gradle:49]: Worth opening a follow up JIRA to track this?

**Review by epugh — COMMENTED** [review]
I haven't tried the code yet, but in reading it appears a lot more legible....

File: solr/solrj/build.gradle (hunk 7 above)
Comment (dsmiley) [reply solr/solrj/build.gradle:49]: Umm; I don't care enough any way.  I did call it out in the PR conversation.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java (hunk 6 above)
Comment (dsmiley) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java]: I'll add a line -- it's rather fundamental to this PR.

For user level, I added a note to solr-upgrade-notes.adoc

**Comment by dsmiley** [comment]
Other than a CHANGES.txt entry, I think this PR is ready.

**Comment by epugh** [comment]
Looks like the content about `CloudSolrClient` in the ref guide (https://solr.apache.org/guide/solr/latest/deployment-guide/solrj.html#building-and-running-solrj-applications)  refers to the use of the solr url for confiugration   I wonder if we should tweak the text thgouh to stress that using ZK is a vulnerability?  Similar to what you put in the upgrade notes?  

Also, do we need to mention the need for `solrj-zookeeper` dependency for `CloudSolrClient` users?

**Comment by epugh** [comment]
Really looking forward to this!

**Comment by epugh** [comment]
Docs look good!

**Comment by epugh** [comment]
:shipit: ???

**Comment by dsmiley** [comment]
Planned commit message after the first line:
```
• CloudSolrClient builders use reflection to instantiate ZkClientClusterStateProvider when needed
• CollectionAdminParams constants
• CollectionStateProps constants
• SliceStateProps constants
• ReplicaStateProps constants
• SolrJ streaming expressions: don’t depend on ZK anymore
• javadocs: temporarily unlink to solrj modules due to split package issue
• PerReplicaStates.fetch -> PerReplicaStatesFetcher
• ZkStateReader.getCollectionPath -> DocCollection.getCollectionPath
```

Some other code changes seemed too obscure to mention.

**Comment by dsmiley** [comment]
I think it's done and I've addressed comments.  I would love to see this merged soon so as to unblock other efforts that touch lots of files.  If nobody has further review comments, I'll merge it Thursday.

File: solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/ZkClientClusterStateProvider.java [hunk 8]
```diff
@@ -67,6 +72,50 @@ public ZkClientClusterStateProvider(String zkHost) {
...
+   * @return the ClusterState
+   */
+  @SuppressWarnings({"unchecked"})
+  @Deprecated
+  public static ClusterState createFromJsonSupportingLegacyConfigName(
+      int version, byte[] bytes, Set<String> liveNodes, String coll, SolrZkClient zkClient) {
+    if (bytes == null || bytes.length == 0) {
+      return new ClusterState(liveNodes, Collections.emptyMap());
+    }
+    Map<String, Object> stateMap = (Map<String, Object>) Utils.fromJSON(bytes);
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/ZkClientClusterStateProvider.java:93]: @dsmiley @heythm is this an issue or can we ignore it?

File: solr/core/src/java/org/apache/solr/logging/MDCLoggingContext.java [hunk 9]
```diff
@@ -108,6 +108,16 @@ private static void setNodeName(String node) {
     }
   }
 
+  public static String getNodeName() {
+    String s = MDC.get(NODE_NAME_PROP);
+    if (s == null) return null;
+    if (s.startsWith("n:")) {
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply solr/core/src/java/org/apache/solr/logging/MDCLoggingContext.java:114]: Just saw that this was copied and pasted from elsewhere, so I guess ignore...

File: solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/ZkClientClusterStateProvider.java [hunk 10]
```diff
@@ -67,7 +67,51 @@ public ZkClientClusterStateProvider(String zkHost) {
...
+     *     Json representation of a {@link DocCollection} as written by {@link ClusterState#write(JSONWriter)}. It
+     *     can represent one or more collections.
+     * @param liveNodes list of live nodes
+     * @param coll collection name
+     * @param zkClient ZK client
+     * @return the ClusterState
+     */
+    @SuppressWarnings({"unchecked"})
+    @Deprecated
+    public static ClusterState createFromJsonSupportingLegacyConfigName(
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/ZkClientClusterStateProvider.java]: That's @janhoy 's PR, I don't have an opinion.

File: solr/solrj-zookeeper/build.gradle [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    testImplementation project(':solr:test-framework')
+    testImplementation project(':solr:core')
+
+    testImplementation 'junit:junit'
+    testImplementation 'commons-io:commons-io'
+    testImplementation 'com.google.guava:guava'
+
+    permitTestUsedUndeclared project(':solr:solrj-zookeeper') // duh!
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply solr/solrj-zookeeper/build.gradle:56]: I see that you just did this to get the build to pass, but it's definitely odd. Not a complaint, just confusion.

**Review by HoustonPutman — APPROVED** [review]
Some small comments and questions, but generally everything looks good.

Thanks for the hard work @heythm and @dsmiley!

File: solr/solrj/build.gradle [hunk 12]
```diff
@@ -50,20 +41,16 @@ dependencies {
...
+  testImplementation project(':solr:solrj-zookeeper')
+  testImplementation 'org.apache.zookeeper:zookeeper'
+  permitTestUnusedDeclared 'org.apache.zookeeper:zookeeper'
```
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/solrj/build.gradle:52]: @risdenk recently added a plugin to our Gradle build that demands we declare all dependencies of a compile time nature, even transitives.  This obviously adds additional lines across our build to maintain.  I wish it was smarter, for example, don't do this for Gradle's rather unique "api".  And it has bugs where we have to add "permitBlahBlahBlah" where it shouldn't be necessary ideally.

File: solr/solrj-zookeeper/build.gradle (hunk 11 above)
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/solrj-zookeeper/build.gradle:56]: Very odd indeed!  See the previous comment I added a minute ago.  CC @risdenk

File: solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/ZkClientClusterStateProvider.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/solrj-zookeeper/src/java/org/apache/solr/client/solrj/impl/ZkClientClusterStateProvider.java:93]: We can't have Utils.getJson depending on ZK.  There was one such method but it was only used in one place so we basically inlined the implementation.  I don't think @heythm wrote this code; it just moved.

I am curious about what Lift's analysis sees; I'll look closer.

File: solr/core/src/java/org/apache/solr/logging/MDCLoggingContext.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/logging/MDCLoggingContext.java:114]: Yes; just moved the method.

File: solr/solrj/build.gradle (hunk 12 above)
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply solr/solrj/build.gradle:52]: Ahhh good to know, thanks for the context!

[the remark this answers is not in the document]
Comment (risdenk) [reply solr/solrj/build.gradle:52]: Hmmm thats interesting - we have very very few of these in the code already.

```
➜  solr git:(SOLR-16187) git grep -F Unused | grep -F build.gradle
solr/core/build.gradle:  permitUnusedDeclared 'com.github.spotbugs:spotbugs-annotations'
solr/modules/hadoop-auth/build.gradle:  permitUnusedDeclared 'com.github.spotbugs:spotbugs-annotations'
solr/modules/hadoop-auth/build.gradle:  permitUnusedDeclared 'org.apache.hadoop:hadoop-annotations'
solr/modules/hdfs/build.gradle:  permitTestUnusedDeclared 'org.apache.zookeeper:zookeeper'
solr/modules/jwt-auth/build.gradle:  //permitTestUnusedDeclared 'com.fasterxml.jackson.core:jackson-databind'
solr/modules/sql/build.gradle:  permitUnusedDeclared 'org.apache.calcite.avatica:avatica-core'
solr/solrj/build.gradle:  permitUnusedDeclared 'com.github.spotbugs:spotbugs-annotations'
solr/test-framework/build.gradle:  permitUnusedDeclared 'com.github.spotbugs:spotbugs-annotations'
solr/test-framework/build.gradle:  permitUnusedDeclared 'commons-cli:commons-cli'
solr/webapp/build.gradle:  permitUnusedDeclared project(":solr:core")
```

Most of these are spotbugs annotations. I know there were bugs in older versions of the plugin. The hdfs module does have the same thing for zookeeper. So I'm not really sure what is going on.

File: solr/solrj-zookeeper/build.gradle (hunk 11 above)
[the remark this answers is not in the document]
Comment (risdenk) [reply solr/solrj-zookeeper/build.gradle:56]: I don't quite understand this one. We only do this for solr/webapp and solr:core. We don't have any existing ones that use `permitTestUsedUndeclared` in the same module where its defined...

[the remark this answers is not in the document]
Comment (risdenk) [reply solr/solrj-zookeeper/build.gradle:56]: We don't use `permitTestUsedUndeclared` anywhere else so something is probably wired up weird here.

File: solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java [hunk 13]
```diff
@@ -115,9 +115,9 @@ public void test() throws Exception {
     log.info("excluded_node : {}  ", emptyNode);
     createReplaceNodeRequest(nodeToBeDecommissioned, emptyNode, null)
         .processAndWait("000", cloudClient, 15);
+    ZkStateReader zkStateReader = ZkStateReader.from(cloudClient);
```
Comment (risdenk) [inline solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java:118]: Does this zkStateReader need to be closed?

File: gradle/documentation/render-javadoc.gradle [hunk 14]
```diff
@@ -325,6 +342,7 @@ class RenderJavadocTask extends DefaultTask {
     // lucene-test-framework was first, or broken links to things like LuceneTestCase if lucene-core was first)
     if (project.path != ':solr:test-framework') {  //
       findRenderTasksInDependencies()
+          .findAll({it.project.path != ':solr:solrj' && it.project.path != ':solr:solrj-zookeeper'}) // split packages
```
Comment (risdenk) [inline gradle/documentation/render-javadoc.gradle:345]: Why is this needed? I know it mentions split packages - but don't we have split packages elsewhere and those modules aren't listed here?

File: solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java (hunk 13 above)
Comment (dsmiley) [reply solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java:118]: No; it was simply extracted from the CloudSolrClient's `ZkClientClusterStateProvider`

File: gradle/documentation/render-javadoc.gradle (hunk 14 above)
Comment (dsmiley) [reply gradle/documentation/render-javadoc.gradle:345]: The link checker of our docs complains about split packages here but perhaps it didn't for other split packages cases?  This is possible; it depends on if the dependency reflects itself in the public API.  If there is a compile dependency that isn't reflected in the API, then there is no javadoc issue.

File: solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java (hunk 13 above)
Comment (HoustonPutman) [reply solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java:118]: The method name does make it sound like it's possibly creating a new instance. Maybe renaming the method to `extractFrom` could be helpful? Not a deal breaker, just could be a bit easier to understand.

Comment (dsmiley) [reply solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java:118]: Note this particular choice was established in SOLR-16061 in March.  Renaming this method would be a separate issue.  `extractFrom` seems fine; perhaps even shorter, just `get`?

Comment (risdenk) [reply solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java:118]: Does it hurt to close the ZkStateReader? Wrap in try-with-resources?

```
try (ZkStateReader zkStateReader = ZkStateReader.from(cloudClient)) {
...
}
```

I'm not that worried about the name, but want to make sure if things are closable they are closed.

Comment (dsmiley) [reply solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java:118]: We shouldn't do that.  `ZkStateReader.from` is a glorified "get" but keeps ZooKeeper dependencies off of CloudSolrClient.

Comment (dsmiley) [reply solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java:118]: Not that SolrClient is already Closable and we should already be taking care to manage their lifecycle.  A CloudSolrClient that uses a ZK state provider will close it.
